googlesource/googlesource.go

```diff
@@ -27,42 +29,64 @@ import (
 	sdk "github.com/conduitio/conduit-connector-sdk"
 	googlebigquery "github.com/neha-Gupta1/conduit-connector-bigquery"
 	"google.golang.org/api/iterator"
-	"google.golang.org/api/option"
 )
 
 var (
 	newClient = bigquery.NewClient
 )
 
 type readRowInput struct {
-	tableID  string
-	position Position
-	wg       *sync.WaitGroup
+	tableID   string
+	offset    string
+	positions map[string]string
+	wg        *sync.WaitGroup
 }
 
+func (s *Source) checkInitialPos(positions map[string]string, incrementColName map[string]string, tableID string, primaryColName map[string]string) (firstSync, userDefinedOffset bool, userDefinedKey bool) {
+	if _, ok := positions[tableID]; !ok {
+		firstSync = true
+	}
+	if _, ok := incrementColName[tableID]; ok {
+		userDefinedOffset = true
+	}
+
+	if _, ok := primaryColName[tableID]; ok {
+		userDefinedKey = true
+	}
+
+	return firstSync, userDefinedOffset, userDefinedKey
+}
+
+// haris: why does rowInput need to be a chan?
+// Neha: the function is getting called inside a goroutine we get wrong value (everytime the last possible values) and
+// func param will change for each function call
 func (s *Source) ReadGoogleRow(rowInput chan readRowInput, responseCh chan sdk.Record) (err error) {
```

Comment:
The function is getting used as- 

	for _, tableID := range s.tables {
		wg.Add(1)

		s.tomb.Go(func() (err error) {
			sdk.Logger(s.ctx).Trace().Msg(fmt.Sprintf("position %v : %v", tableID, s.getPosition()[tableID]))
			return s.ReadGoogleRow(rowInput, s.records)
		})
		position := s.getPosition()
		rowInput <- readRowInput{tableID: tableID, offset: position[tableID], positions: position, wg: &wg}
	}

Since we are running it in goroutine we were required to add tomb to handle errors. `tomb.Go` takes input in format of  `s.tomb.Go(func() (err error)`. We added a closure function to provide the function in the required format.
But we were required to pass the `tableID` and other values. So one way to do this is by passing values through a channel. This will make sure that on every call of a function inside the goroutine we pass the correct value maintaining the format required by tomb.

@hariso We can also have a call if required to discuss this.

Reply:
Actually, if we do a single table per source, then we won't need a tomb to coordinate multiple goroutines at all.

Reply:
Yes. If we have just one table all this will go.